:bug: Skip HelmChartProxy reconciliation for paused clusters

jimmidyson opened the pull request:
This fixes a bug leading to duplicate `HelmReleaseProxy` resources
following a `clusterctl move` that can happen due to a race between
reconciling the `HelmChartProxy` and moving and unpausing the `Cluster`.

Fixes #188.

jimmidyson commented:
Tested locally and this fixed my issue btw.

mboersma reviewed (commented):
Thanks for this! It does seem like a fix.

Would you be up for writing a unit test that captures this behavior in `helmchartproxy_controller_test.go`? It should be mostly copy-and-paste to add a new Cluster definition with `Spec.Paused` == true, then a new test case that ensures the HCP is Ready (and that it isn't ready without your change).

mboersma commented:
/cc @Jont828

jimmidyson commented:
@mboersma On it! Thanks for your review!

Jont828 commented on controllers/helmchartproxy/helmchartproxy_controller.go:
```diff
@@ -218,6 +218,12 @@ func (r *HelmChartProxyReconciler) reconcileNormal(ctx context.Context, helmChar
 		}
 
+		// Don't reconcile if the Cluster is paused.
+		if cluster.Spec.Paused {
```
```suggestion
		if annotations.IsPaused(cluster, cluster) {
```
Seems like the convention in CAPI is to use this annotation getter.

jimmidyson replied:
Ah good tip! Just a small adjustment to check the paused annotation on the `helmChartProxy` and the `cluster.Spec.Paused` via the call you suggest. I'm also going to move this into `reconcileForCluster` which I think makes more sense.

Jont828 commented on controllers/helmchartproxy/helmchartproxy_controller.go:
```diff
@@ -218,6 +218,12 @@ func (r *HelmChartProxyReconciler) reconcileNormal(ctx context.Context, helmChar
 
+		// Don't reconcile if the Cluster is paused.
+		if cluster.Spec.Paused {
+			log.V(2).Info("Cluster is paused, skipping reconcile", "cluster", cluster.Name)
```
```suggestion
			log.V(2).Info("Reconciliation is paused for this Cluster", "cluster", cluster.Name)
```

jimmidyson replied:
Done, thanks!

Jont828 reviewed (commented):
Small nits, but otherwise a pretty straightforward change.

jimmidyson commented:
Added test cases, tested the HRP gets created without this change, and with this change does not. Also added test case to ensure when a cluster is unpaused the HRP gets created.

mboersma commented:
/assign @Jont828

Jont828 commented:
Please squash your commits, otherwise looks good!

jimmidyson commented:
@Jont828 @mboersma Squashed! Please lgtm approve when you get a moment 🙏 

And if you could cut a release once this is merged that would be most appreciated, it's blocking some of our e2e tests.

mboersma reviewed (approved):
/lgtm

jimmidyson commented:
/assign @Jont828

jackfrancis reviewed (approved):
/lgtm
/approve

mboersma commented:
@jimmidyson unfortunately @Jont828 is away for several days, and the other maintainers aren't familiar with the release process. 

We have a bus factor problem here. I'll create an issue for the lack of release documentation. (If you or someone you know would like to be a reviewer or maintainer on CAAPH, I think we would entertain that idea too.)

jimmidyson commented:
@mboersma Thanks for the update. I'd happily help out as a maintainer if you'll have me 😅

mboersma commented:
Thanks @jimmidyson! Let's wait until later next week when the current maintainers are all available and I'll propose it.

Jont828 commented:
@jimmidyson Hey, I'm back from vacation. Super glad to hear that you're interested in getting involved! I'll reach out on Slack if you'd like to chat more.
